src/pythium/histogramming/objects.py

```diff
@@ -0,0 +1,371 @@
+        else:
+            raise TypeError("CrossProduct object can be accessed by integers or one of following keys (sample, region, obserbale, systematic, template)")
+    
+    def __iter__(self):
+        return iter(self._xp)
+
+
+
+class Observable(object):
+    '''
+    This class defines an observable that will be retrieved from all samples entering a
+    given region, and constructed with the given binning. 
+    
+    Attributes:
+        name (str): 
+            The name given to the observable 
+        var (str):
+            The name of the observable in the input file
+        binning (:py:class:`pythium.histogramming.binning._Binning`):
+            The chosen binning for this observable
+        dataset (str):
+            The equivalent of a TTree in ROOT files. It is the parent group for
+            the observable in the input file (e.g. nominal tree in a ROOT file)
+        obs_build ( :py:class:`pythium.common.functor.Functor` ):
+            An instance of :py:class:`pythium.common.functor.Functor` which defines
+            how to build the observable from existing data
+    '''
+
+    TObservable = TypeVar("TObservable", bound="Observable")
+    T = TypeVar("T")
+    TypeOrListOfTypes = Union[T, List[T]]
+    ListOrListOfLists = Union[List[T], List[List[T]]]
+    NumbersArray = Union[np.ndarray, List[Union[int, float]]]
+
+    @beartype
+    def __init__( 
+        self, var: TypeOrListOfTypes[str], 
+        name: str, 
+        binning: TypeOrListOfTypes[_Binning], 
+        dataset: str, 
+        weights: TypeOrListOfTypes[Union[str, NumbersArray]] = 1.,
+        label: Optional[str] = '', 
+        samples: Optional[List[str]] = None, 
+        exclude_samples: Optional[List[str]] = None,
+        regions: Optional[List[str]] = None,
+        exclude_regions: Optional[List[str]] = None, 
+        *,
+        obs_build: Optional[TypeOrListOfTypes[Functor]] = None
+    ) -> None :
+        
+        # Everything arrays to support ndim operations
+        self.var = var if isinstance(var, list) else [var,]
+        self.binning = binning if isinstance(binning, list) else [binning]
+        self.axes = self.get_axes()
+        self.weights = weights if isinstance(weights, list) else [weights]*len(self.var)
+        if isinstance(weights, (int, float)):   self.weights = weights
+        h_attr = [self.var, self.binning, self.axes, self.weights]
+        assert all(len(attr) == len(h_attr[0]) for attr in h_attr if (attr != self.weights and weights!=1.))
+        self.ndim = len(h_attr[0])
+
+        self.builder = obs_build
+        if self.ndim != 1 and self.builder is not None:  logger.error("Building n-dim observables on the fly is not supported")
+        if self.builder is None:   self.builder = Functor(lambda *args: args, self.var, reqvars= self.var, new = False)
+       
+       
+       
+        self.name = name
```

Comment:
Move to line 99 (before self.binning = ...). Can you remember why?

Reply:
I can't remember why at all. Can you maybe run with the wrong order to see where the error is thrown?